v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/transform/AccumulatingTableReader.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (C) 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.source.reader.io.transform;
+
+import com.google.auto.value.AutoValue;
+import com.google.cloud.teleport.v2.source.reader.io.row.SourceRow;
+import com.google.cloud.teleport.v2.source.reader.io.schema.SourceTableReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.Map.Entry;
+import org.apache.beam.sdk.transforms.Count;
+import org.apache.beam.sdk.transforms.Flatten;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+
+/**
+ * Helper to {@link ReaderTransform} that expands the Pipeline Graph to flatten all the tableReaders
+ * and generate table completions. Please refer to {@link ReaderTransform} for greater details.
+ */
+@AutoValue
+abstract class AccumulatingTableReader extends PTransform<PBegin, PCollectionTuple> {
```

Comment:
This does not account for reading tables in a DAG right? This call will change when we add that support?

Reply:
Yup, but the sequencing can be added here with a easy interace. It is not implemented to avoid complexity till it is required.

For example:
Currently AccumulatingTableReader has a `List<PTransforms>`
That can be changed to a `List<List<Ptransforms>>`
The external code, will call
with_table_reader()
with_table_reader()
and_then()
with_table_reader()

At each `and_then` we will append to a new list and before falttening, we will add `WaitOn` tranforming between outer lists.

This is currently not implemented to avoid un-necessary complexity.

Reply:
Thanks for the detailed discussion.
As discussed I will add a note on sequencing and signals (used to potentially trigger event for live pipeline)  in the design.